src/SlackFunction.ts

```diff
@@ -46,83 +113,412 @@ export class SlackFunction {
+   *
+   * @param args
+   * @returns A `complete()` utility callback
+   * which can be accessed from any SlackFunction
+   * handler and used to to complete your SlackFunction.
+   *
+   * ```
+   * Example:
+   * const handler = async ({ complete }) => { complete() };
+   * const myFunc = new SlackFunction("id", handler);
+   * ```
+   */
+  // eslint-disable-next-line
+  private createComplete(args: any): CompleteFunction {
+    const { payload, body, client } = args;
+
+    // gets the function execution id from a function executed event
+    let { function_execution_id } = payload;
+
+    // gets the function execution id from the function data in a function interactivity event
+    if (function_execution_id === undefined && body !== undefined && 'function_data' in body) {
+      function_execution_id = body.function_data.execution_id;
     }
+
+    // if stil undefined, error
+    if (function_execution_id === undefined) {
+      const msg = '⚠️ Cannot generate required complete utility without a function_execution_id';
+      throw new SlackFunctionCompleteError(msg);
+    }
+
+    // return the utility callback
+    return ({ outputs, error }: CompleteFunctionArgs = {}) => {
+      // Slack API requires functions complete with either outputs or error, not both
+      if (outputs && error) {
+        throw new SlackFunctionCompleteError('⚠️ Cannot complete with outputs and error message supplied');
+      }
+      // if user has supplied outputs OR has supplied neither outputs nor error
+      if (outputs !== undefined || (outputs === undefined && error === undefined)) {
+        // helpful logging
+        this.logger.debug('🚀 Attempting to complete with outputs:', outputs);
+        return client.apiCall('functions.completeSuccess', {
+          outputs: outputs ?? {},
+          function_execution_id,
+        });
+      } if (error !== undefined) {
+        this.logger.debug('🚀 Attempting to complete with error:', error);
+        return client.apiCall('functions.completeError', {
+          error,
+          function_execution_id,
+        });
+      }
+      return null;
+    };
+  }
+}
+
+/* Event handling validation */
+
+export function isFunctionExecutedEvent(args: AnyMiddlewareArgs): boolean {
+  if (args.payload === undefined) {
     return false;
   }
+  return (args.payload.type === 'function_executed');
+}
 
-  private run = async (args: AnyMiddlewareArgs & AllMiddlewareArgs): Promise<void> => {
-    const fnArgs = prepareFnArgs(args);
-    this.fn(fnArgs);
-  };
+export function isFunctionInteractivityEvent(args: AnyMiddlewareArgs & AllMiddlewareArgs): boolean {
+  const allowedInteractivityTypes = [
+    'block_actions', 'view_submission', 'view_closed'];
+  if (args.body === undefined) return false;
+  return (
+    allowedInteractivityTypes.includes(args.body.type) &&
+    ('function_data' in args.body)
+  );
 }
 
-function isFunctionExecutedEvent(args: AnyMiddlewareArgs): boolean {
```

Comment:
You've added export to all these private functions but is it necessary? Do we use those outside this source file?

Reply:
`passInteractivityConstraints` in particular, no it seems isn't used, so I can remove this export for now. 

`isFunctionInteractivityEvent` `isFunctionExecutedEvent` `hasHandler` `hasCallbackId` `passConstraint` are being imported in unit tests.

Reply:
I see. That makes sense!